runtime/caml/skiplist.h

```diff
@@ -0,0 +1,102 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*             Xavier Leroy, projet Cambium, INRIA Paris                  */
+/*                                                                        */
+/*   Copyright 2020 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+/* A dictionary data structure implemented as skip lists */
+
+/* Keys and associated data are natural-width integers (type [uintnat]).
+   Pointers can be used too, modulo conversion to [uintnat]. */
+
+#ifndef CAML_SKIPLIST_H
+#define CAML_SKIPLIST_H
+
+#include "config.h"
+
+#define NUM_LEVELS 17
+
+/* The head of a skip list */
+
+struct skiplist {
```

Comment:
I haven't quite made up my mind.  Basically I'm following the pattern set by `struct ext_table` in `<caml/misc.h>`:
- it's not `CAML_INTERNALS` because extensible tables and skip lists can legitimately be used by third-party C/OCaml stub code.
- Functions are prefixed with `caml_` because it's a linking nightmare otherwise.
- Types and macros are not prefixed because it doesn't seem to be a problem so far.

Reply:
When you added `struct ext_table` in 2001, that predates a lot of the concern for prefixing, etc. (it predates the `caml_` prefixing in 2004).

All said, the header isn't (yet) included by any other header, so a decision isn't needed until it "leaks". Might it be worth therefore protecting the whole of `skiplist.h` in a `CAML_INTERNALS`, just so that that's consciously done in future if skiplists start appearing in the public API? I have future #8896-like problems in mind.

Reply:
(NB if `skiplist.h` is protected by `CAML_INTERNALS`, then `skiplist.c` will need the customary `#define CAML_INTERNALS` at the top - the other two already have it)

Reply:
Since you are very keen on this name spacing issue, I added protection by `CAML_INTERNALS`.